File: global_scheduler/brute_force_solver.py
import sys
import os
import logging
from typing import List, Callable, Dict
from copy import deepcopy
from more_itertools import distinct_permutations
from global_scheduler.structs import Job
from global_scheduler.simulator import WeaveSimulator

current_dir = os.path.dirname(os.path.abspath(__file__))
build_dir = os.path.join(current_dir, 'build')
sys.path.insert(0, build_dir)


class IntraGroupSolver:
    def __init__(self, jobs: List[Job], sim_steps: int = 1000):
        self.jobs = jobs
        self.sim_steps = sim_steps
        try:
            import global_scheduler_cpp
            cpp_jobs = []
            for job in jobs:
                cpp_job = global_scheduler_cpp.Job(
                    job.job_id,
                    job.t_rollout,
                    job.t_train,
                    job.rollout_nodes or [],
                    job.train_nodes or []
                )
                cpp_jobs.append(cpp_job)
            self.cpp_solver = global_scheduler_cpp.IntraGroupSolver(cpp_jobs, sim_steps)
            logging.info("[IntraGroupSolver] Using C++ backend...")
        except:
            self.cpp_solver = None
            logging.warning("[IntraGroupSolver] C++ backend is not available, fallback to native python backend.")

    def _generate_deployment(self):
        # We assume t_rollout >= t_train, thus we only consider independent
        # rollout (i.e., N rollout and only 1 train)
        # Also, assume no job will simultaneously use multiple rollout nodes
        def gen_partitions(jobs):
            # Find all possible ways to partition jobs into rollout nodes
            if len(jobs) == 0:
                return [[]]
            first, *rest = jobs
            partitions = gen_partitions(rest)
            result = []
            for part in partitions:
                # Put `first` in a new subset
                result.append([[first]] + part)
                # Put `first` into each existing subset
                for i in range(len(part)):
                    new_part = [subset[:] for subset in part]
                    new_part[i].append(first)
                    result.append(new_part)
            return result

        all_parts = gen_partitions(list(range(len(self.jobs))))
        # Normalize: sort each subset and sort list of subsets
        normalized = []
        seen = set()
        for part in all_parts:
            sorted_boxes = [sorted(box) for box in part]
            sorted_boxes.sort()  # sort boxes lexicographically
            key = tuple(tuple(box) for box in sorted_boxes)
            if key not in seen:
                seen.add(key)
                normalized.append(sorted_boxes)
        
        return normalized

    def _convert_cpp_jobs_to_python(self, cpp_jobs):
        python_jobs = []
        for cpp_job in cpp_jobs:
            python_job = type('Job', (), {})()
            python_job.job_id = cpp_job.job_id
            python_job.t_rollout = cpp_job.t_rollout
            python_job.t_train = cpp_job.t_train
            python_job.rollout_nodes = cpp_job.rollout_nodes
            python_job.train_nodes = cpp_job.train_nodes
            python_jobs.append(python_job)
        return python_jobs

    def solve(self, max_meta_iter_len: int, score_func: Callable[[Dict], float]):
        if self.cpp_solver is not None:
            def cpp_score_func(rollout_utils: List[float], train_utils: List[float]) -> float:
                utils_dict = {'rollout': rollout_utils, 'train': train_utils}
                return score_func(utils_dict)
            result = self.cpp_solver.solve(max_meta_iter_len, cpp_score_func)
            best_strategy = (
                result.partition,
                self._convert_cpp_jobs_to_python(result.job_deployment),
                result.meta_iteration
            )
            return result.score, best_strategy
        all_partitions = self._generate_deployment()
        idx_2_job_id = [job.job_id for job in self.jobs]

        def assemble_jobs(partition: List[List[int]]):
            ret_jobs = deepcopy(self.jobs)
            for rollout_node_id, group in enumerate(partition):
                for job_id in group:
                    assert ret_jobs[job_id].rollout_nodes is None
                    ret_jobs[job_id].rollout_nodes = [f'RN-{rollout_node_id}']
                    ret_jobs[job_id].train_nodes = ['TN']
            return ret_jobs

        def generate_compositions(n: int, k: int, min_val: int):
            # Genetate all ways to a number n into k parts [n_1,...,n_k],
            # subject to sum_{j=1}^{k}n_j = n and \forall j, n_j >= min_val
            if k == 1:
                if n >= min_val:
                    yield [n]
                return
            for i in range(min_val, n - min_val * (k - 1) + 1):
                for rest in generate_compositions(n - i, k - 1, min_val):
                    yield [i] + rest
        
        def generate_meta_iteration_all():
            assert max_meta_iter_len >= len(self.jobs)
            all_combs = []
            for meta_iter_len in range(len(self.jobs), max_meta_iter_len + 1):
                for composition in generate_compositions(meta_iter_len, len(self.jobs), 1):
                    meta_iter_comb = []
                    for i in range(len(composition)):
                        meta_iter_comb += [idx_2_job_id[i]] * composition[i]
                    all_combs.append(meta_iter_comb)
            all_meta_iters = []
            for job_comb in all_combs:
                all_meta_iters += list(distinct_permutations(job_comb))
            return all_meta_iters

        meta_iterations = generate_meta_iteration_all()
        max_score, best_strategy = -float("inf"), None
        print(f"Generated {len(all_partitions)} partitions and {len(meta_iterations)} meta iterations")
        for partition in all_partitions:
            job_deployment = assemble_jobs(partition)
            for meta_iter in meta_iterations:
                sim = WeaveSimulator(job_deployment, meta_iter)
                _, _, utils = sim.simulate_run(self.sim_steps)
                score = score_func(utils)
                if score > max_score:
                    max_score = score
                    best_strategy = (partition, job_deployment, meta_iter)
        return max_score, best_strategy

File: global_scheduler/csrc/brute_force_solver.cpp
#include "brute_force_solver.hpp"
#include <iostream>
#include <unordered_map>
#include <set>

IntraGroupSolver::IntraGroupSolver(const std::vector<Job>& jobs, int sim_steps)
    : original_jobs_(jobs), sim_steps_(sim_steps) {
    
    // Build index to job_id mapping
    for (const auto& job : original_jobs_) {
        idx_2_job_id_.push_back(job.job_id);
    }
}

// Partition generation using iterative approach for better performance
std::vector<std::vector<std::vector<int>>> IntraGroupSolver::generatePartitions(int n) {
    std::vector<std::vector<std::vector<int>>> all_partitions;
    
    if (n == 0) {
        all_partitions.push_back({});
        return all_partitions;
    }
    
    // Initialize with first element
    std::vector<std::vector<std::vector<int>>> partitions = {{{0}}};
    
    // Add elements one by one
    for (int i = 1; i < n; ++i) {
        std::vector<std::vector<std::vector<int>>> new_partitions;
        
        for (const auto& partition : partitions) {
            // Option 1: put element i in a new subset
            auto new_partition1 = partition;
            new_partition1.push_back({i});
            new_partitions.push_back(new_partition1);
            
            // Option 2: put element i in each existing subset
            for (size_t j = 0; j < partition.size(); ++j) {
                auto new_partition2 = partition;
                new_partition2[j].push_back(i);
                new_partitions.push_back(new_partition2);
            }
        }
        
        partitions = std::move(new_partitions);
    }
    
    // Normalize partitions (sort subsets and sort list of subsets)
    std::set<std::vector<std::vector<int>>> normalized_set;
    
    for (auto& partition : partitions) {
        // Sort each subset
        for (auto& subset : partition) {
            std::sort(subset.begin(), subset.end());
        }
        // Sort list of subsets lexicographically
        std::sort(partition.begin(), partition.end());
        normalized_set.insert(partition);
    }
    
    all_partitions.assign(normalized_set.begin(), normalized_set.end());
    return all_partitions;
}

// Composition generation using iterative approach
std::vector<std::vector<int>> IntraGroupSolver::generateCompositions(int n, int k, int min_val) {
    std::vector<std::vector<int>> result;
    std::vector<int> current;
    generateCompositionsHelper(n, k, min_val, current, result);
    return result;
}

void IntraGroupSolver::generateCompositionsHelper(int n, int k, int min_val, 
                                                std::vector<int>& current, 
                                                std::vector<std::vector<int>>& result) {
    if (k == 1) {
        if (n >= min_val) {
            current.push_back(n);
            result.push_back(current);
            current.pop_back();
        }
        return;
    }
    
    for (int i = min_val; i <= n - min_val * (k - 1); ++i) {
        current.push_back(i);
        generateCompositionsHelper(n - i, k - 1, min_val, current, result);
        current.pop_back();
    }
}

// Distinct permutations generation
std::vector<std::vector<std::string>> IntraGroupSolver::generateDistinctPermutations(
    const std::vector<std::string>& elements) {
    
    std::vector<std::vector<std::string>> result;
    if (elements.empty()) return result;
    
    std::vector<std::string> mutable_elements = elements;
    generateDistinctPermutationsHelper(mutable_elements, 0, result);
    return result;
}

void IntraGroupSolver::generateDistinctPermutationsHelper(
    std::vector<std::string>& elements, int start, 
    std::vector<std::vector<std::string>>& result) {
    
    if (start == elements.size() - 1) {
        result.push_back(elements);
        return;
    }
    
    std::unordered_set<std::string> seen;
    for (int i = start; i < elements.size(); ++i) {
        if (seen.find(elements[i]) != seen.end()) continue;
        seen.insert(elements[i]);
        
        std::swap(elements[start], elements[i]);
        generateDistinctPermutationsHelper(elements, start + 1, result);
        std::swap(elements[start], elements[i]);
    }
}

// Meta iteration generation
std::vector<std::vector<std::string>> IntraGroupSolver::generateMetaIterations(int max_meta_iter_len) {
    int n_jobs = original_jobs_.size();
    std::vector<std::vector<std::string>> all_meta_iters;
    
    for (int meta_iter_len = n_jobs; meta_iter_len <= max_meta_iter_len; ++meta_iter_len) {
        auto compositions = generateCompositions(meta_iter_len, n_jobs, 1);
        
        for (const auto& composition : compositions) {
            std::vector<std::string> job_comb;
            for (int i = 0; i < n_jobs; ++i) {
                for (int j = 0; j < composition[i]; ++j) {
                    job_comb.push_back(idx_2_job_id_[i]);
                }
            }
            
            auto permutations = generateDistinctPermutations(job_comb);
            all_meta_iters.insert(all_meta_iters.end(), permutations.begin(), permutations.end());
        }
    }
    
    return all_meta_iters;
}

// Job assembly
std::vector<Job> IntraGroupSolver::assembleJobs(const std::vector<std::vector<int>>& partition) {
    std::vector<Job> ret_jobs = original_jobs_;
    
    for (size_t rollout_node_id = 0; rollout_node_id < partition.size(); ++rollout_node_id) {
        for (int job_idx : partition[rollout_node_id]) {
            ret_jobs[job_idx].rollout_nodes = {"RN-" + std::to_string(rollout_node_id)};
            ret_jobs[job_idx].train_nodes = {"TN"};
        }
    }
    
    return ret_jobs;
}

// Main solver
Solution IntraGroupSolver::solve(int max_meta_iter_len, 
                               const std::function<double(const std::vector<double>&, 
                                                        const std::vector<double>&)>& score_func) {
    
    if (max_meta_iter_len < original_jobs_.size()) {
        throw std::invalid_argument("max_meta_iter_len must be >= number of jobs");
    }
    
    auto all_partitions = generatePartitions(original_jobs_.size());
    auto meta_iterations = generateMetaIterations(max_meta_iter_len);
    
    std::cout << "Generated " << all_partitions.size() << " partitions and " 
              << meta_iterations.size() << " meta iterations" << std::endl;
    
    Solution best_solution;
    
    int counter = 0;
    for (const auto& partition : all_partitions) {
        auto job_deployment = assembleJobs(partition);
        
        for (const auto& meta_iter : meta_iterations) {
            WeaveSimulator sim(job_deployment, meta_iter);
            auto result = sim.simulate_run(sim_steps_);
            
            double score = score_func(result.rollout_utils, result.train_utils);
            
            if (++counter % 1000 == 0) {
                std::cout << "Processed " << counter << " combinations, current best score: " 
                          << best_solution.score << std::endl;
            }
            
            if (score > best_solution.score) {
                best_solution.score = score;
                best_solution.partition = partition;
                best_solution.job_deployment = job_deployment;
                best_solution.meta_iteration = meta_iter;
            }
        }
    }
    
    std::cout << "Best score: " << best_solution.score << std::endl;
    return best_solution;
}